examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php

```diff
@@ -0,0 +1,602 @@
+        // The limit to the number of listing group filters permitted in a Performance
+        // Max campaign can be found here:
+        // https://developers.google.com/google-ads/api/docs/best-practices/system-limits.
+        foreach ($response->iterateAllElements() as $googleAdsRow) {
+            /** @var GoogleAdsRow $googleAdsRow */
+            $assetGroupListingGroupFilters[] = $googleAdsRow->getAssetGroupListingGroupFilter();
+        }
+        return $assetGroupListingGroupFilters;
+    }
+    // [END add_performance_max_product_listing_group_tree_7]
+
+    /**
+     * Removes an existing tree of asset group listing group filters.
+     *
+     * Asset group listing group filters must be removed in a specific order: all of the children
+     * of a filter must be removed before the filter itself, otherwise the API will return an
+     * error.
+     *
+     * @param GoogleAdsClient $googleAdsClient the Google Ads API client
+     * @param int $customerId the customer ID
+     * @param AssetGroupListingGroupFilter[] $assetGroupListingGroupFilters the existing asset
+     *     group listing group filters
+     * @return MutateOperation[] the list of MutateOperations to remove all listing groups
+     */
+    private static function createMutateOperationsForRemovingListingGroupFiltersTree(
+        array $assetGroupListingGroupFilters
+    ): array {
+        if (empty($assetGroupListingGroupFilters)) {
+            throw new \UnexpectedValueException('No listing group filters to remove');
+        }
+
+        $resourcesToListingGroupFilters = [];
+        $parentsToChildren = [];
+        $rootResourceName = null;
+        foreach ($assetGroupListingGroupFilters as $assetGroupListingGroupFilter) {
+            $resourcesToListingGroupFilters[$assetGroupListingGroupFilter->getResourceName()] =
+                $assetGroupListingGroupFilter;
+            // When the node has no parent, it means it's the root node, which is treated
+            // differently.
+            if (empty($assetGroupListingGroupFilter->getParentListingGroupFilter())) {
+                if (!is_null($rootResourceName)) {
+                    throw new \UnexpectedValueException('More than one root node found.');
+                }
+                $rootResourceName = $assetGroupListingGroupFilter->getResourceName();
+                continue;
+            }
+
+            $parentResourceName = $assetGroupListingGroupFilter->getParentListingGroupFilter();
+            $siblings = [];
+
+            // Checks to see if we've already visited a sibling in this group, and fetches or
+            // creates a new set as required.
+            if (array_key_exists($parentResourceName, $parentsToChildren)) {
+                $siblings = $parentsToChildren[$parentResourceName];
+            }
+            $siblings[] = $assetGroupListingGroupFilter->getResourceName();
+            $parentsToChildren[$parentResourceName] = $siblings;
```

Comment:
This is only necessary when executed in an `else` clause of the condition above

Reply:
I don't think so. Can you tell why you think like that?

Reply:
It works but I find the logic a bit confusing.

I would expect that we only have to execute `$parentsToChildren[$parentResourceName] = $siblings` with `$siblings` a new array when `$parentsToChildren[$parentResourceName]` is unset.

I guess is it nit-picking so we can keep it as-is.